### PR TITLE
fix(logging): do not log exceptions twice, deprecate `traceback_line_limit` and fix `pretty_print_tty`

### LINKED ISSUE #3228: Bug: exceptions are logged twice
### Description

Exceptions are logged twice

### URL to code causing the issue

_No response_

### MCVE

```python
from litestar import get, Litestar, Request
from litestar.logging import LoggingConfig


@get("/exc")
async def exception_route(request: Request) -> str:
    request.logger.info("Reaching exception route endpoint")
    raise RuntimeError("Something went wrong")
    return "never reached"


app = Litestar(route_handlers=[exception_route],
               logging_config=LoggingConfig(log_exceptions="always"))
```

### Steps to reproduce

_No response_

### Screenshots

_No response_

### Logs

```
$ .venv/bin/litestar run
Using Litestar app from app:app
Starting server process ─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
┌──────────────────────────────┬──────────────────────┐
│ Litestar version             │ 2.7.0                │
│ Debug mode                   │ Disabled             │
│ Python Debugger on exception │ Disabled             │
│ CORS                         │ Disabled             │
│ CSRF                         │ Disabled             │
│ OpenAPI                      │ Enabled path=/schema │
│ Compression                  │ Disabled             │
└──────────────────────────────┴──────────────────────┘
INFO:     Started server process [94846]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
INFO - 2024-03-19 12:48:04,135 - root - app - Reaching exception route endpoint
INFO:     127.0.0.1:58369 - "GET /exc HTTP/1.1" 500 Internal Server Error
ERROR - 2024-03-19 12:48:04,136 - litestar - config - exception raised on http connection to route /exc

Traceback (most recent call last):
  File "/Users/jderrien/projects/litestar-minimal/.venv/lib/python3.11/site-packages/litestar/middleware/exceptions/middleware.py", line 218, in __call__
    await self.app(scope, receive, send)
  File "/Users/jderrien/projects/litestar-minimal/.venv/lib/python3.11/site-packages/litestar/routes/http.py", line 82, in handle
    response = await self._get_response_for_request(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jderrien/projects/litestar-minimal/.venv/lib/python3.11/site-packages/litestar/routes/http.py", line 134, in _get_response_for_request
    return await self._call_handler_function(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jderrien/projects/litestar-minimal/.venv/lib/python3.11/site-packages/litestar/routes/http.py", line 154, in _call_handler_function
    response_data, cleanup_group = await self._get_response_data(
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jderrien/projects/litestar-minimal/.venv/lib/python3.11/site-packages/litestar/routes/http.py", line 207, in _get_response_data
    data = await route_handler.fn(**parsed_kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jderrien/projects/litestar-minimal/app.py", line 8, in exception_route
    raise RuntimeError("Something went wrong")
RuntimeError: Something went wrong
Traceback (most recent call last):
  File "/Users/jderrien/projects/litestar-minimal/.venv/lib/python3.11/site-packages/litestar/middleware/exceptions/middleware.py", line 218, in __call__
    await self.app(scope, receive, send)
  File "/Users/jderrien/projects/litestar-minimal/.venv/lib/python3.11/site-packages/litestar/routes/http.py", line 82, in handle
    response = await self._get_response_for_request(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jderrien/projects/litestar-minimal/.venv/lib/python3.11/site-packages/litestar/routes/http.py", line 134, in _get_response_for_request
    return await self._call_handler_function(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jderrien/projects/litestar-minimal/.venv/lib/python3.11/site-packages/litestar/routes/http.py", line 154, in _call_handler_function
    response_data, cleanup_group = await self._get_response_data(
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jderrien/projects/litestar-minimal/.venv/lib/python3.11/site-packages/litestar/routes/http.py", line 207, in _get_response_data
    data = await route_handler.fn(**parsed_kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jderrien/projects/litestar-minimal/app.py", line 8, in exception_route
    raise RuntimeError("Something went wrong")
RuntimeError: Something went wrong
```

### Litestar Version

litestar==2.7.0

### Platform

- [ ] Linux
- [X] Mac
- [ ] Windows
- [ ] Other (Please specify in the description above)

---
> [!NOTE]  
> While we are open for sponsoring on [GitHub Sponsors](https://github.com/sponsors/litestar-org/) and 
> [OpenCollective](https://opencollective.com/litestar), we also utilize [Polar.sh](https://polar.sh/) to engage in pledge-based sponsorship.
>
> Check out all issues funded or available for funding [on our Polar.sh dashboard](https://polar.sh/litestar-org)
> * If you would like to see an issue prioritized, make a pledge towards it!
> * We receive the pledge once the issue is completed & verified
> * This, along with engagement in the community, helps us know which features are a priority to our users.

<a href="https://polar.sh/litestar-org/litestar/issues/3228">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/litestar-org/litestar/issues/3228/pledge.svg?darkmode=1">
  Fund with Polar
</picture>
</a>

### PR DESCRIPTION
**Pull request by jderrien**
## Description

Currently we're logging the exceptions twice: the full one and a truncated one.
This PR tries to improve this, ~~which implies small breaking changes~~.

~~Almost finished but still a WIP, I need some feedback please!~~

Outlines:

* The wording of the log message, when logging an exception, has been updated.
* For structlog, the `traceback` field in the log message (which contained a truncated stacktrace) has been removed. The `exception` field is still around and contains the full stacktrace.
* The option `traceback_line_limit` has been deprecated. The value is now ignored, the full stacktrace will be logged.

### Summary

⚠️ The former code has been introduced to solve #1294, unfortunately I haven't been able to reproduce the issue. It's not clear in which cases it happened, or if the PR (#1296) fixed the original issue (I don't think so because exception are logged twice in the end).
❓ **TODO** Anyway, to solve the original issue (#1294), truncating the request's body would have been a better solution. I'd be happy to dig a little bit more here. Any idea about how to reproduce the issue? (I already tried to write a test with a file upload, but the request's body doesn't get printed)
⚠️ ~~Type `ExceptionLoggingHandler` and signature of `handle_exception_logging()` get updated by this PR, seems better and more flexible this way.~~

#### Logging
~~✅ Behavior is still the same, just slightly changed how we count the lines to display (see the code and the test `test_traceback_truncate_default_logging`).~~
~~✅ Ability to set `traceback_line_limit` to `None`.~~
~~⚠️ Default value for `traceback_line_limit` is `None`~~, I don't think we should truncate stacktraces by default.
❓ → ✅ Maybe we could deprecate `traceback_line_limit` to keep things in sync with Structlog. And if it's really something wanted by users, they can bring it back easily via a custom `exception_logging_handler`.
ℹ️ Updated log messages to get something similar to structlog.

#### StructLog
✅ Fixed a bug with `pretty_print_tty` (as a separated commit).
✅ When `pretty_print_tty=True`, we don't log the stacktrace twice and the remaining one gets truncated automatically by structlog (see below).
✅ When `pretty_print_tty=False`, exception gets logged as json. Previously it was logged in 2 separate fields (`exception` contains the full stacktrace & `traceback` a truncated version). This PR removes `traceback`. It will reduce the log size. It's still possible to parse the JSON after, with some tools, to limit the lines displayed.
❓ → ✅ If it's fine on your side, can I mark `traceback_line_limit` as deprecated and document it?

#### Extras
⚠️ Note: A few tests use `structlog.testing.capture_logs`. When using it to capture the `logging.exception()` output, the traceback is not yet rendered. That's something to keep in mind.
❓ What about turning `log_exceptions` to `always` by default? (Hiding errors by default seems a bad choice according to my sysadmin background).

#### With `pretty_print_tty=False`

Before, both `traceback` and `exception` were logged:
```
{
  "connection_type": "http",
  "path": "/test",
  "traceback": "  File \"/Users/jderrien/projects/litestar/litestar/middleware/_internal/exceptions/middleware.py\", line 158, in __call__\n    await self.app(scope, receive, capture_response_started)\n  File \"/Users/jderrien/projects/litestar/litestar/_asgi/asgi_router.py\", line 99, in __call__\n    await asgi_app(scope, receive, send)\n  File \"/Users/jderrien/projects/litestar/litestar/routes/http.py\", line 80, in handle\n    response = await self._get_response_for_request(\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/Users/jderrien/projects/litestar/litestar/routes/http.py\", line 132, in _get_response_for_request\n    return await self._call_handler_function(\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/Users/jderrien/projects/litestar/litestar/routes/http.py\", line 152, in _call_handler_function\n    response_data, cleanup_group = await self._get_response_data(\n                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/Users/jderrien/projects/litestar/litestar/routes/http.py\", line 203, in _get_response_data\n    data = route_handler.fn(**parsed_kwargs)\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/Users/jderrien/projects/litestar/tests/unit/test_middleware/test_exception_handler_middleware.py\", line 299, in handler\n    raise ValueError(\"Test debug exception\")\nValueError: Test debug exception\n",
  "event": "Uncaught Exception",
  "level": "error",
  "exception": "Traceback (most recent call last):\n  File \"/Users/jderrien/projects/litestar/litestar/middleware/_internal/exceptions/middleware.py\", line 158, in __call__\n    await self.app(scope, receive, capture_response_started)\n  File \"/Users/jderrien/projects/litestar/litestar/_asgi/asgi_router.py\", line 99, in __call__\n    await asgi_app(scope, receive, send)\n  File \"/Users/jderrien/projects/litestar/litestar/routes/http.py\", line 80, in handle\n    response = await self._get_response_for_request(\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/Users/jderrien/projects/litestar/litestar/routes/http.py\", line 132, in _get_response_for_request\n    return await self._call_handler_function(\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/Users/jderrien/projects/litestar/litestar/routes/http.py\", line 152, in _call_handler_function\n    response_data, cleanup_group = await self._get_response_data(\n                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/Users/jderrien/projects/litestar/litestar/routes/http.py\", line 203, in _get_response_data\n    data = route_handler.fn(**parsed_kwargs)\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/Users/jderrien/projects/litestar/tests/unit/test_middleware/test_exception_handler_middleware.py\", line 299, in handler\n    raise ValueError(\"Test debug exception\")\nValueError: Test debug exception",
  "timestamp": "2024-05-16T22:07:01.991184Z"
}
```

After, we only log `exception` and the global size is reduced:
```
{
  "connection_type": "http",
  "path": "/test",
  "event": "Uncaught exception",
  "level": "error",
  "exception": "Traceback (most recent call last):\n  File \"/Users/jderrien/projects/litestar/litestar/middleware/_internal/exceptions/middleware.py\", line 156, in __call__\n    await self.app(scope, receive, capture_response_started)\n  File \"/Users/jderrien/projects/litestar/litestar/_asgi/asgi_router.py\", line 99, in __call__\n    await asgi_app(scope, receive, send)\n  File \"/Users/jderrien/projects/litestar/litestar/routes/http.py\", line 80, in handle\n    response = await self._get_response_for_request(\n               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/Users/jderrien/projects/litestar/litestar/routes/http.py\", line 132, in _get_response_for_request\n    return await self._call_handler_function(\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/Users/jderrien/projects/litestar/litestar/routes/http.py\", line 152, in _call_handler_function\n    response_data, cleanup_group = await self._get_response_data(\n                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/Users/jderrien/projects/litestar/litestar/routes/http.py\", line 203, in _get_response_data\n    data = route_handler.fn(**parsed_kwargs)\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/Users/jderrien/projects/litestar/tests/unit/test_middleware/test_exception_handler_middleware.py\", line 310, in handler\n    raise ValueError(\"Test debug exception\")\nValueError: Test debug exception",
  "timestamp": "2024-05-16T22:15:12.091675Z"
}
```

#### With `pretty_print_tty=True`:

Before, the truncated `traceback` value doesn't really help.
```
2024-05-16T22:03:40.912722Z [error    ] Uncaught Exception             connection_type=http path=/test traceback=  File "/Users/jderrien/projects/litestar/litestar/middleware/_internal/exceptions/middleware.py", line 158, in __call__
    await self.app(scope, receive, capture_response_started)
  File "/Users/jderrien/projects/litestar/litestar/_asgi/asgi_router.py", line 99, in __call__
    await asgi_app(scope, receive, send)
  File "/Users/jderrien/projects/litestar/litestar/routes/http.py", line 80, in handle
    response = await self._get_response_for_request(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jderrien/projects/litestar/litestar/routes/http.py", line 132, in _get_response_for_request
    return await self._call_handler_function(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jderrien/projects/litestar/litestar/routes/http.py", line 152, in _call_handler_function
    response_data, cleanup_group = await self._get_response_data(
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jderrien/projects/litestar/litestar/routes/http.py", line 203, in _get_response_data
    data = route_handler.fn(**parsed_kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jderrien/projects/litestar/tests/unit/test_middleware/test_exception_handler_middleware.py", line 299, in handler
    raise ValueError("Test debug exception")
ValueError: Test debug exception

╭───────────────────── Traceback (most recent call last) ──────────────────────╮
│ /Users/jderrien/projects/litestar/litestar/middleware/_internal/exceptions/m │
│ iddleware.py:158 in __call__                                                 │
│                                                                              │
│   155 │   │   │   await send(event)                                          │
│   156 │   │                                                                  │
│   157 │   │   try:                                                           │
│ ❱ 158 │   │   │   await self.app(scope, receive, capture_response_started)   │
│   159 │   │   except Exception as e:                                         │
│   160 │   │   │   if scope_state.response_started:                           │
│   161 │   │   │   │   raise LitestarException("Exception caught after respon │
│                                                                              │
│ /Users/jderrien/projects/litestar/litestar/_asgi/asgi_router.py:99 in        │
│ __call__                                                                     │
│                                                                              │
│    96 │   │   else:                                                          │
│    97 │   │   │   ScopeState.from_scope(scope).exception_handlers = route_ha │
│    98 │   │   │   scope["route_handler"] = route_handler                     │
│ ❱  99 │   │   await asgi_app(scope, receive, send)                           │
│   100 │                                                                      │
│   101 │   @lru_cache(1024)  # noqa: B019                                     │
│   102 │   def handle_routing(self, path: str, method: Method | None) -> tupl │
│                                                                              │
│                           ... 3 frames hidden ...                            │
│                                                                              │
│ /Users/jderrien/projects/litestar/litestar/routes/http.py:203 in             │
│ _get_response_data                                                           │
│                                                                              │
│   200 │   │   │   │   │   else await route_handler.fn(**parsed_kwargs)       │
│   201 │   │   │   │   )                                                      │
│   202 │   │   elif route_handler.has_sync_callable:                          │
│ ❱ 203 │   │   │   data = route_handler.fn(**parsed_kwargs)                   │
│   204 │   │   else:                                                          │
│   205 │   │   │   data = await route_handler.fn(**parsed_kwargs)             │
│   206                                                                        │
│                                                                              │
│ /Users/jderrien/projects/litestar/tests/unit/test_middleware/test_exception_ │
│ handler_middleware.py:299 in handler                                         │
│                                                                              │
│   296 def test_traceback_truncate_struct_logging() -> None:                  │
│   297 │   @get("/test")                                                      │
│   298 │   def handler() -> None:                                             │
│ ❱ 299 │   │   raise ValueError("Test debug exception")                       │
│   300 │                                                                      │
│   301 │   app = Litestar([handler], logging_config=StructLoggingConfig(log_e │
│   302                                                                        │
╰──────────────────────────────────────────────────────────────────────────────╯
ValueError: Test debug exception
```

After (more concise):
```
2024-05-16T22:17:36.360643Z [error    ] Uncaught exception             connection_type=http path=/test
╭───────────────────── Traceback (most recent call last) ──────────────────────╮
│ /Users/jderrien/projects/litestar/litestar/middleware/_internal/exceptions/m │
│ iddleware.py:156 in __call__                                                 │
│                                                                              │
│   153 │   │   │   await send(event)                                          │
│   154 │   │                                                                  │
│   155 │   │   try:                                                           │
│ ❱ 156 │   │   │   await self.app(scope, receive, capture_response_started)   │
│   157 │   │   except Exception as e:                                         │
│   158 │   │   │   if scope_state.response_started:                           │
│   159 │   │   │   │   raise LitestarException("Exception caught after respon │
│                                                                              │
│ /Users/jderrien/projects/litestar/litestar/_asgi/asgi_router.py:99 in        │
│ __call__                                                                     │
│                                                                              │
│    96 │   │   else:                                                          │
│    97 │   │   │   ScopeState.from_scope(scope).exception_handlers = route_ha │
│    98 │   │   │   scope["route_handler"] = route_handler                     │
│ ❱  99 │   │   await asgi_app(scope, receive, send)                           │
│   100 │                                                                      │
│   101 │   @lru_cache(1024)  # noqa: B019                                     │
│   102 │   def handle_routing(self, path: str, method: Method | None) -> tupl │
│                                                                              │
│                           ... 3 frames hidden ...                            │
│                                                                              │
│ /Users/jderrien/projects/litestar/litestar/routes/http.py:203 in             │
│ _get_response_data                                                           │
│                                                                              │
│   200 │   │   │   │   │   else await route_handler.fn(**parsed_kwargs)       │
│   201 │   │   │   │   )                                                      │
│   202 │   │   elif route_handler.has_sync_callable:                          │
│ ❱ 203 │   │   │   data = route_handler.fn(**parsed_kwargs)                   │
│   204 │   │   else:                                                          │
│   205 │   │   │   data = await route_handler.fn(**parsed_kwargs)             │
│   206                                                                        │
│                                                                              │
│ /Users/jderrien/projects/litestar/tests/unit/test_middleware/test_exception_ │
│ handler_middleware.py:310 in handler                                         │
│                                                                              │
│   307 def test_traceback_truncate_struct_logging() -> None:                  │
│   308 │   @get("/test")                                                      │
│   309 │   def handler() -> None:                                             │
│ ❱ 310 │   │   raise ValueError("Test debug exception")                       │
│   311 │                                                                      │
│   312 │   app = Litestar([handler], logging_config=StructLoggingConfig(log_e │
│   313                                                                        │
╰──────────────────────────────────────────────────────────────────────────────╯
ValueError: Test debug exception
```

## Closes

Closes #3228

### CODE REVIEW

**Comment by peterschutt** [comment]
> which implies small breaking changes.

Can you ELI5 these please?

**Comment by jderrien** [comment]
@peterschutt Basically, that's what I tried to do with the ⚠️ icons. It also depends what you define as a breaking change.

To sum up:
* Update of type `ExceptionLoggingHandler` and signature of `handle_exception_logging()` might break custom `exception_logging_handler`. This is the main breaking change.
* Update of log messages might break some tooling.
* For structlog, the field `traceback` is removed. It might break some tooling too.
* Adding `None` as a possible value to `traceback_line_limit`, and also as the default value, might break something.

**Comment by peterschutt** [comment]
> @peterschutt Basically, that's what I tried to do with the ⚠️ icons. It also depends what you define as a breaking change.
> 
> To sum up:
> 
>     * Update of type `ExceptionLoggingHandler` and signature of `handle_exception_logging()` might break custom `exception_logging_handler`. This is the main breaking change.
> 
>     * Update of log messages might break some tooling.
> 
>     * For structlog, the field `traceback` is removed. It might break some tooling too.
> 
>     * Adding `None` as a possible value to `traceback_line_limit`, and also as the default value, might break something.

Thanks for that. I think this means you'd need to point this at the v3 branch, unless you think there's a way you can do this in a non-breaking way.

However, on the topic of pointing this as v3, you should also read https://github.com/litestar-org/litestar/issues/2858

**Comment by jderrien** [comment]
@peterschutt Thank you for pointing me to this issue. Is there a timeline for a v3 realase?

In fact, I have this PR but I also started working on other improvements on another branch a while ago. So I'm wondering if it's worth to continue the work. If v2 is still around for some months (like 4+ months), I think it might be worth.

> I think this means you'd need to point this at the v3 branch, unless you think there's a way you can do this in a non-breaking way.

I can probably rollback the change on `ExceptionLoggingHandler` / `handle_exception_logging()`. But the other changes are problematic too? I mean, having `traceback` for structlog doesn't make sense since `exception` is logged. It's just increasing the log size for nothing.

My take: When you discover a new framework, you generally start a new project and one of the first tasks is to setup the logging. The current state of litestar logging doesn't inspire trust in the framework. So it might be interesting to solve what can be easily solved, maybe at the cost of minor breaking changes. And if v3 can get rid of the logging complexity in the future, it would be even better.

**Comment by peterschutt** [comment]
> @peterschutt Thank you for pointing me to this issue. Is there a timeline for a v3 realase?

No timeline at this point.

> In fact, I have this PR but I also started working on other improvements on another branch a while ago. So I'm wondering if it's worth to continue the work. If v2 is still around for some months (like 4+ months), I think it might be worth.

V2 will still be around for a long time, we'll continue to support it just as we currently support V1 (aka starlite). V2 will go into maintenance mode once V3 is out.

> > I think this means you'd need to point this at the v3 branch, unless you think there's a way you can do this in a non-breaking way.
> 
> I can probably rollback the change on `ExceptionLoggingHandler` / `handle_exception_logging()`. But the other changes are problematic too? I mean, having `traceback` for structlog doesn't make sense since `exception` is logged. It's just increasing the log size for nothing.

> My take: When you discover a new framework, you generally start a new project and one of the first tasks is to setup the logging. The current state of litestar logging doesn't inspire trust in the framework. So it might be interesting to solve what can be easily solved, maybe at the cost of minor breaking changes. And if v3 can get rid of the logging complexity in the future, it would be even better.

Perhaps we should bite the bullet and deprecate the use of LoggingConfig in the next minor release? And at the same time, replace the usage documentation for logging with examples of how to configure popular logging libraries alongside litestar. It would be a step in the direction of https://github.com/litestar-org/litestar/issues/2858#issuecomment-2076300407, without yet making changes that would break existing code.

**Comment by jderrien** [comment]
> Perhaps we should bite the bullet and deprecate the use of LoggingConfig in the next minor release? And at the same time, replace the usage documentation for logging with examples of how to configure popular logging libraries alongside litestar. It would be a step in the direction of [#2858 (comment)](https://github.com/litestar-org/litestar/issues/2858#issuecomment-2076300407), without yet making changes that would break existing code.

Yep, I read https://github.com/litestar-org/litestar/issues/2858#issuecomment-2076300407 and that's what I was refering to in my previous message. I already had the same feeling.

But I feel like deprecating `LoggingConfig`right now is more work for both maintainers and users. So if it was me, I would probably keep that for v3.

I think that what I call "small breaking changes" here are easy to adapt for the minority of users impacted (who should be familiar with `LoggingConfig`). And it improves things for everyone (decrease log size, exceptions more readable, small bug fixes, etc).

Despite of the complexity behind `LoggingConfig`, it's customizable enough. Overall it does the job, so it could be kept around in v2. But new users shouldn't have to dive into this complexity at the beginning of their journey with Litestar.

When I discovered Litestar and started a project a few weeks ago, with Python 3.12, the `queue_listener` handler was broken, the exceptions are not displayed by default, I had to read the docs, search for github issues, dig into the litestar codebase to understand what was going on. And then I worked on #3185. It was interesting but also quite hard and time consuming.

According to #3482 I'm not the only one.

So my main goal is just to get decent logging defaults out of the box for new users. I fixed the `queue_listener` handler in #3185. There is this PR. And then I plan to open another one (see this branch: https://github.com/jderrien/litestar/tree/logging-improvements).

Depending on your final decision, I could continue this work and we could make a single release with all these logging breaking changes (but trying to limit them). Or I could stop the work around this subject.

Just let me know :)

**Comment by peterschutt** [comment]
> > Perhaps we should bite the bullet and deprecate the use of LoggingConfig in the next minor release? And at the same time, replace the usage documentation for logging with examples of how to configure popular logging libraries alongside litestar. It would be a step in the direction of [#2858 (comment)](https://github.com/litestar-org/litestar/issues/2858#issuecomment-2076300407), without yet making changes that would break existing code.
> 
> Yep, I read [#2858 (comment)](https://github.com/litestar-org/litestar/issues/2858#issuecomment-2076300407) and that's what I was refering to in my previous message. I already had the same feeling.
> 
> But I feel like deprecating `LoggingConfig`right now is more work for both maintainers and users. So if it was me, I would probably keep that for v3.

If we are to make the proposed changes in v3, the current behavior _must_ be deprecated in v2 - the functionality won't go away, we just wouldn't be suggesting its use.

> When I discovered Litestar and started a project a few weeks ago, with Python 3.12, the `queue_listener` handler was broken, the exceptions are not displayed by default, I had to read the docs, search for github issues, dig into the litestar codebase to understand what was going on. Then I discovered exceptions were logged twice and the framework was trying to limit the size of the stacktraces. Finally I worked on #3185. It was interesting but also quite hard and time consuming.

To me, this is another indicator that this facet of our code base should be deprecated, and removed. If so many surface-level issues are present, how much could the feature possibly be in use in the wild? I for one don't use it.

> According to #3482 I'm not the only one to struggle.
> 
> So my main goal is just to get decent logging defaults out of the box for new users. I fixed the `queue_listener` handler in #3185. There is this PR. And then I plan to open another one (see this branch: https://github.com/jderrien/litestar/tree/logging-improvements).
> 
> Depending on your final decision, I could continue this work and we could make a single release with all these logging breaking changes (but trying to limit them). Or I could stop the work around this subject.
> 
> Just let me know :)

@litestar-org/maintainers - any input here?

**Comment by jderrien** [comment]
@provinzkraut While we agree on this, the question now is about this PR and the future one I might open.

To sum up, `LoggingConfig` will have to stay here for a while, even if you decide to deprecate it, as @peterschutt said. It will take time before being able to remove it completly. So would you accept my fixes at the cost of minor breaking changes? My opinion is it will benefit to new users and improve things for existing ones.

Here are the breaking changes I have mentioned for this PR in a previous comment, with more context (in bold):
* Update of type `ExceptionLoggingHandler` and signature of `handle_exception_logging()` might break custom `exception_logging_handler`. This is the main breaking change. **This might be reverted**
* Update of log messages might break some tooling. **This improves readability & `logging` gets more aligned with `structlog`**
* For structlog, the field `traceback` is removed. It might break some tooling too. **But it will primarily decrease log size (important in production) and improves readability**
* Adding `None` as a possible value to `traceback_line_limit`, and also as the default value, might break something. **The signature gets updated. Tracebacks are no more truncated by default, which doesn't make sense, knowing we currently have a full traceback and a possibly truncated one. For `structlog`, this option would be deprecated.**

**Comment by jderrien** [comment]
Hey @provinzkraut,

I've update the PR. No more breaking changes in the API.

---

About removing `traceback`:

> Is this necessary? From what I understand, this just tidies things up a bit.

I'd say this is necessary since #3228 is "exceptions are logged twice". So we have to remove this field for structlog to fix the issue. The full stacktrace was and is still logged in the `exception` field.

AFAIK, it might break some external tool setup (i.e. log parsing). But this is almost the same risk as updating the wording of log messages. So not really a breaking change (?).

In the end, this is easy to update for users. I think the release note should just make it clear.

The outcome is:
* Reduce the log size for everyone, which is important in production environments
* Improve the developper experience (logs are easier to read in development, and new users don't have to dig too much into the logging to get a proper config)

I don't think we should wait v3 release to fix that.

---

About adding `None` as a possible value to `traceback_line_limit`:

I've reverted it. And I wasn't comfortable with keeping the feature in the standard logging config while removing it in the structlog one.
So default value is now `-1` in both cases, and any other value is ignored and will show a deprecation warning.

I'm wondering if a deprecation warning is enough (since they're masked by default)?

---

For a possible release, if I don't miss anything, the outline might be:

Exception logging:
* The wording of the log message has been updated
* For `structlog`, the `traceback` field in the log message (which contained a truncated stacktrace) has been removed. The `exception` field already contains the full stacktrace.
* The option `traceback_line_limit` has been deprecated, the value is now ignored, the full stacktrace will be logged

File: litestar/logging/config.py
Comment (JacobCoffee) [inline litestar/logging/config.py:147]:
```suggestion
    .. deprecated:: 2.9.0
    	This parameter is deprecated and ignored. It will be removed in a future release.
```

File: litestar/logging/config.py
Comment (JacobCoffee) [inline litestar/logging/config.py:225]:
```suggestion
    .. deprecated:: 2.9.0
    	This parameter is deprecated and ignored. It will be removed in a future release.
```

File: litestar/logging/config.py
Comment (JacobCoffee) [inline litestar/logging/config.py:444]:
```suggestion
    .. deprecated:: 2.9.0
    	This parameter is deprecated and ignored. It will be removed in a future release.
```

**Review by JacobCoffee — APPROVED** [review]

**Review by provinzkraut — APPROVED** [review]

**Comment by provinzkraut** [comment]
Thanks for the effort @jderrien!

This is good to go in from my side. The coverage decrease is expected since you're deleting tests, so that can be ignored!

**Review by provinzkraut — APPROVED** [review]

**Pull request merged**
